### PR TITLE
PNP rule

### PR DESCRIPTION
**Pull request by danjuv**
This PR attempts to add PnP support to the `yarn_install` rule in bazel. PnP is designed to replace the `node_modules` folder with a lookup table pointing directly to the yarn cache..

The core idea behind the design is to emulate the way PnP resolving works by providing the integrity checksum as input to bazel instead of individual files. Which means we save time normally spent symlinking node_modules into the sandbox/repository directories since we only need to pass around a solitary .pnp.js file (the file containing the lookup table).

The main drawback of this design is the pnp file references the yarn cache, which isn't sandboxed by bazel (and we don't want it to, since that would be more time consuming than fine-grained node_modules). While bazel cannot guarantee the hermeticity of the yarn cache, determinism is a core feature of yarn, so the rules as a whole will still be deterministic.

### CODE REVIEW

File: internal/node/node.bzl [hunk 1]
```diff
@@ -239,6 +239,7 @@ def _nodejs_binary_impl(ctx):
         "TEMPLATED_repository_args": _to_manifest_path(ctx, ctx.file._repository_args),
         "TEMPLATED_script_path": _to_execroot_path(ctx, ctx.file.entry_point),
         "TEMPLATED_vendored_node": "" if is_builtin else strip_external(ctx.file._node.path),
+        "TEMPLATED_pnp_file": _to_manifest_path(ctx, ctx.file.pnp_file),
```
Comment (uri-canva) [inline internal/node/node.bzl]: This doesn't need to be its own substitution: if we use this we don't use `TEMPLATED_bazel_require_script`, so you can replace that with this when `pnp_file` is set.

File: internal/node/node_launcher.sh [hunk 2]
```diff
@@ -207,7 +206,7 @@ if [ "${EXPECTED_EXIT_CODE}" -eq "0" ]; then
   # handled by the node process.
   # If we had merely forked a child process here, we'd be responsible
   # for forwarding those OS interactions.
-  exec "${node}" "${NODE_OPTIONS[@]}" "${MAIN}" "${ARGS[@]}"
+  exec "${node}" "${ARGS[@]}" "${NODE_OPTIONS[@]}" "$(pwd)/TEMPLATED_script_path"
```
Comment (uri-canva) [inline internal/node/node_launcher.sh]: Why the change from `MAIN` to `$(pwd)/TEMPLATED_script_path`? Does it not work with the relative path? If so why not change the value of `MAIN`?

File: internal/npm_install/generate_pnp_build_files.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  fs.writeFileSync(p, content);
+}
+
+function createBuildFile(pkg: string) {
+  const contents = `package(default_visibility = ["//visibility:public"])
+load("@build_bazel_rules_nodejs//internal/npm_install:node_module_library.bzl", "node_module_library")
+
+filegroup(
+name = "${pkg}",
+srcs = []
```
Comment (uri-canva) [inline internal/npm_install/generate_pnp_build_files.ts]: Add a comment here explaining that the `srcs` are intentionally empty, otherwise it would look like a mistake.

File: internal/npm_install/generate_pnp_build_files.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  const flattenedDeps = await processDependencies(dependencies, "dependencies");
+
+  flattenedDeps.forEach(createBuildFile);
+
+  processDependencies(devDependencies, "devDependencies");
+
+  const contents = `package(default_visibility = ["//visibility:public"])
+exports_files([
+  ".pnp.js",
+  "package.json",
```
Comment (uri-canva) [inline internal/npm_install/generate_pnp_build_files.ts]: 4 space indent in `BUILD.bazel` files.

File: internal/npm_install/generate_pnp_build_files.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  mkdirp(path.dirname(p));
+  fs.writeFileSync(p, content);
+}
+
+function createBuildFile(pkg: string) {
+  const contents = `package(default_visibility = ["//visibility:public"])
+load("@build_bazel_rules_nodejs//internal/npm_install:node_module_library.bzl", "node_module_library")
+
+filegroup(
+name = "${pkg}",
```
Comment (uri-canva) [inline internal/npm_install/generate_pnp_build_files.ts]: 4 space indent in `BUILD.bazel` files.

File: internal/npm_install/generate_pnp_build_files.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  writeFileSync(`${pkg}/BUILD.bazel`, contents);
+}
+
+async function processDependencies(
+  deps: string[],
+  type: "dependencies" | "devDependencies"
+): Promise<Set<string>> {
+  const flattenedDeps: Set<string> = new Set();
+  deps.forEach((key: string) => {
+    // loop over transitive dependencies here
```
Comment (uri-canva) [inline internal/npm_install/generate_pnp_build_files.ts]: Remove this comment, it's not relevant anymore.

File: internal/npm_install/generate_pnp_build_files.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+async function main() {
+  const packageJson = require(path.join(workspacePath, "package.json"));
+  const dependencies = Object.keys(packageJson.dependencies || {});
+  const devDependencies = Object.keys(packageJson.dependencies || {});
+  const flattenedDeps = await processDependencies(dependencies, "dependencies");
+
+  flattenedDeps.forEach(createBuildFile);
+
+  processDependencies(devDependencies, "devDependencies");
```
Comment (uri-canva) [inline internal/npm_install/generate_pnp_build_files.ts]: We should create build files for dev dependencies too. Are there also test dependencies or other dependencies types we should handle?

File: internal/npm_install/npm_install.bzl [hunk 8]
```diff
@@ -130,22 +130,31 @@ data attribute.
         default = True,
     ),
 })
+YARN_ENVIRONMENT = [
 
-def _create_build_files(repository_ctx, rule_type, node, lock_file):
+]
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl:143]: This is empty? Is this intentional?

File: internal/npm_install/process_pnp_file.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Main entrypoint.
+ */
+function main() {
+  const workspacePath = process.argv.slice(2)[0];
+
+  const pnpPath = path.join(workspacePath, ".pnp.js")
+  const pnpFile = fs.readFileSync(pnpPath, "utf8");
+  const patchedPnpFile = pnpFile.replace("issuerModule ? issuerModule.filename : `${process.cwd()}/`;", `"${process.cwd()}"`);
```
Comment (uri-canva) [inline internal/npm_install/process_pnp_file.ts]: Add a comment explaining the intended replacement here. Otherwise it's very confusing.

**Review by uri-canva — CHANGES_REQUESTED** [review]
@joscha can you review the typescript code to make sure it's idiomatic?
@toby does this make sense to you?

**Review by uri-canva — CHANGES_REQUESTED** [review]
Add an explanation of the design of the change in the PR description. Copy that in the commit message too, so it will show up in the patch when you create it with `git format-patch`.

File: internal/npm_install/npm_install.bzl [hunk 10]
```diff
@@ -130,22 +130,31 @@ data attribute.
...
         repository_ctx.file("manual_build_file_contents", repository_ctx.attr.manual_build_file_contents)
     result = repository_ctx.execute([
         node,
-        "generate_build_file.js",
-        repository_ctx.attr.name,
-        rule_type,
-        "1" if error_on_build_files else "0",
-        repository_ctx.path(lock_file),
-        ",".join(repository_ctx.attr.included_files),
-    ])
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl]: This should be changed to execute the new generator when using pnp, but still running the old one when not.

File: internal/npm_install/npm_install.bzl [hunk 11]
```diff
@@ -343,13 +357,33 @@ def _yarn_install_impl(repository_ctx):
...
+        fail("deleting preinstall scripts failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (result.stdout, result.stderr))
+    result = repository_ctx.execute([
+        "sed",
+        "-i",
+        "-e",
+        "/^ *\"postinstall\".*$/d",
+        repository_ctx.path(repository_ctx.attr.package_json.package + "/package.json"),
+    ])
+    if result.return_code:
+        fail("deleting postinstall scripts failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (result.stdout, result.stderr))
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl:413]: Why deleting the install scripts? Did this come from the other patch?

File: internal/node/node.bzl (hunk 1 above)
Comment (danjuv) [reply internal/node/node.bzl]: I think we'd need to if we're going to keep this backwards compatible though

File: internal/node/node_launcher.sh (hunk 2 above)
Comment (danjuv) [reply internal/node/node_launcher.sh]: because `MAIN` is a node script that patches the resolver to fix the resolve order for node_modules for some cases (like Angular). Since we don't have node modules anymore I replaced that with just invoking the entry script directly.

We can change the value of MAIN though, that will also work

**Review by uri-canva — CHANGES_REQUESTED** [review]

File: internal/npm_install/npm_install.bzl [hunk 12]
```diff
@@ -343,13 +357,33 @@ def _yarn_install_impl(repository_ctx):
...
 
     args = [
         repository_ctx.path(yarn),
         "--cwd",
         root,
         "--network-timeout",
         str(repository_ctx.attr.network_timeout * 1000),  # in ms
+        "--modules-folder",
+        repository_ctx.path("node_modules"),
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl]: What is this change for? Did it come from another patch?

**Review by uri-canva — CHANGES_REQUESTED** [review]

File: internal/npm_install/npm_install.bzl (hunk 11 above)
Comment (danjuv) [reply internal/npm_install/npm_install.bzl:413]: yes, this code came from the yarn_install patch in the canva repo

File: internal/node/node.bzl (hunk 1 above)
Comment (danjuv) [reply internal/node/node.bzl]: I think I'd prefer if it has its own substitution, it makes it easier in the launcher script to verify if we are in PnP mode or not. WDYT?

**Comment by danjuv** [comment]
@uri-canva changes in

File: internal/node/node.bzl [hunk 13]
```diff
@@ -225,12 +225,12 @@ def _nodejs_binary_impl(ctx):
         node_tool_files.append(ctx.outputs.templated_args_file)
 
     is_builtin = ctx.attr._node.label.workspace_name in ["nodejs_%s" % p for p in BUILT_IN_NODE_PLATFORMS]
-
     substitutions = {
         "TEMPLATED_args": " ".join([
             expand_location_into_runfiles(ctx, a)
             for a in templated_args
         ]),
+
```
Comment (uri-canva) [inline internal/node/node.bzl]: Spurious change, revert: the smaller the change the easier it will be to apply cleanly when we update the baseline.

File: internal/node/node.bzl [hunk 14]
```diff
@@ -258,15 +259,15 @@ def _nodejs_binary_impl(ctx):
     # entry point is only needed in runfiles if it is a .js file
     if ctx.file.entry_point.extension == "js":
         runfiles = depset([ctx.file.entry_point], transitive = [runfiles])
-
+    run_files = [ctx.outputs.loader]
```
Comment (uri-canva) [inline internal/node/node.bzl]: In this context there's now a variable called `runfiles` and one called `run_files`. That is too confusing.

File: internal/node/node.bzl [hunk 15]
```diff
@@ -258,15 +259,15 @@ def _nodejs_binary_impl(ctx):
...
+        run_files.append(ctx.file.pnp_file) 
     return [
         DefaultInfo(
             executable = executable,
             runfiles = ctx.runfiles(
                 transitive_files = runfiles,
-                files = node_tool_files + [
-                            ctx.outputs.loader,
-                        ] + ctx.files._source_map_support_files +
+                files = node_tool_files + run_files + ctx.files._source_map_support_files +
```
Comment (uri-canva) [inline internal/node/node.bzl]: Why are `node_tool_files`, `node_modules` and `sources` added to both `files` and `transitive_files`? Does `pnp_file` also need added to both? Or do neither?

File: internal/node/node_launcher.sh [hunk 16]
```diff
@@ -179,13 +179,17 @@ source $repository_args
...
+  NODE_OPTIONS=("--require" "${pnp_file}")
+fi
+
 ALL_ARGS=(TEMPLATED_args $NODE_REPOSITORY_ARGS "$@")
 for ARG in "${ALL_ARGS[@]}"; do
   case "$ARG" in
     --bazel_node_modules_manifest=*) MODULES_MANIFEST="${ARG#--bazel_node_modules_manifest=}" ;;
     --nobazel_patch_module_resolver)
       MAIN="TEMPLATED_script_path"
-      NODE_OPTIONS+=( "--require" "$bazel_require_script" )
```
Comment (uri-canva) [inline internal/node/node_launcher.sh:188]: Restore this to stay backwards compatible.

File: internal/node/node_launcher.sh [hunk 17]
```diff
@@ -179,13 +179,17 @@ source $repository_args
...
+if [[ -n "${pnp_file}" ]]; then
+  MAIN="TEMPLATED_script_path"
+  NODE_OPTIONS=("--require" "${pnp_file}")
+fi
+
 ALL_ARGS=(TEMPLATED_args $NODE_REPOSITORY_ARGS "$@")
 for ARG in "${ALL_ARGS[@]}"; do
   case "$ARG" in
     --bazel_node_modules_manifest=*) MODULES_MANIFEST="${ARG#--bazel_node_modules_manifest=}" ;;
     --nobazel_patch_module_resolver)
```
Comment (uri-canva) [inline internal/node/node_launcher.sh]: I see this is conditional on passing `--nobazel_patch_module_resolver` from https://github.com/bazelbuild/rules_nodejs/blob/5b4e4a0fa403af0aa201ee94e00e9807daee6dbd/internal/providers/node_runtime_deps_info.bzl#L62 can our code do the same, rather than looking for `pnp_file`?

File: internal/npm_install/npm_install.bzl [hunk 18]
```diff
@@ -130,6 +130,39 @@ data attribute.
...
 })
+YARN_ENVIRONMENT = [
+    "CC",
+    "CFLAGS",
+    "CXX",
+    "CXXFLAGS",
+    "JAVA_HOME",
+    "LDFLAGS",
+    "PATH",
+]
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl:143]: Rather than including the changes in the patches we already use, create a new branch with them, and rebase this branch on that branch, so the PR only shows changes in the current patch, not all patches together.

File: internal/npm_install/npm_install.bzl [hunk 19]
```diff
@@ -155,17 +188,26 @@ def _add_scripts(repository_ctx):
         repository_ctx.path(Label("//internal/npm_install:pre_process_package_json.js")),
         {},
     )
-
-    repository_ctx.template(
+    
+    repository_ctx.template(
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl]: Spurious change, revert (it might require editing the file in an editor that doesn't strip trailing whitespace from lines).

File: internal/npm_install/npm_install.bzl [hunk 20]
```diff
@@ -267,6 +309,7 @@ cd "{root}" && "{npm}" {npm_args}
     )
     if result.return_code:
         fail("pre_process_package_json.js failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (result.stdout, result.stderr))
+    is_pnp = result.stdout.strip()
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl]: It looks like there is a json parsing library for starlark: https://github.com/bazelbuild/bazel/issues/3732#issuecomment-502735610

Please use that instead of this, this is too hacky.

**Review by uri-canva — CHANGES_REQUESTED** [review]

File: internal/node/node.bzl (hunk 15 above)
Comment (danjuv) [reply internal/node/node.bzl]: not sure, i tried to change as little as possible, since even if node_modules lists are passed around, they're going to be empty

File: internal/npm_install/generate_pnp_build_files.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+  const contents = `package(default_visibility = ["//visibility:public"])
+exports_files([
+    ".pnp.js",
+    "package.json",
+])`;
+  ``;
+  writeFileSync(`BUILD.bazel`, contents);
+}
+
+main();
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]: inject `process.argv` here:

```suggestion
main(process.argv[3]);
```

so that the file can be required without side-effects or potentical errors.

File: internal/npm_install/generate_pnp_build_files.ts [hunk 22]
```diff
@@ -0,0 +1,44 @@
+import * as fs from "fs";
+import * as path from "path";
+
+const workspacePath = process.argv.slice(2)[0];
+
+function mkdirp(p: string) {
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]: Node 10.12 has `fs.mkdir(..., { recursive: true })` - do we need to support older versions?

File: internal/npm_install/generate_pnp_build_files.ts [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+filegroup(
+    name = "${pkg}",
+    srcs = [] # we dont care about the actual content, we want a list of dependencies to modify the pnp file around
+)`;
+
+  writeFileSync(`${pkg}/BUILD.bazel`, contents);
+}
+
+async function main() {
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]:
```suggestion
async function main(workspacePath: string) {
```

File: internal/npm_install/generate_pnp_build_files.ts [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+const workspacePath = process.argv.slice(2)[0];
+
+function mkdirp(p: string) {
+  if (!fs.existsSync(p)) {
+    mkdirp(path.dirname(p));
+    fs.mkdirSync(p);
+  }
+}
+
+function writeFileSync(p: string, content: string) {
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]: OOI: do all these need to be synchronous?

Also nit: `(p: string, content: string)` is a much more narrow signature than the original `writeFileSync` on `fs`. You could either inline the mkdirp calls (there are only two invocations anyway) or make your own `writeFileSync` a `typeof fs.writeFileSync` to align the signatures. I'd probably do the former.

File: internal/npm_install/generate_pnp_build_files.ts [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+    srcs = [] # we dont care about the actual content, we want a list of dependencies to modify the pnp file around
+)`;
+
+  writeFileSync(`${pkg}/BUILD.bazel`, contents);
+}
+
+async function main() {
+  const packageJson = require(path.join(workspacePath, "package.json"));
+  const dependencies = Object.keys(packageJson.dependencies || {});
+  const devDependencies = Object.keys(packageJson.dependencies || {});
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]:
```suggestion
  const devDependencies = Object.keys(packageJson.devDependencies || {});
```

also: what about peer and optional dependencies?

File: internal/npm_install/generate_pnp_build_files.ts [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+filegroup(
+    name = "${pkg}",
+    srcs = [] # we dont care about the actual content, we want a list of dependencies to modify the pnp file around
+)`;
+
+  writeFileSync(`${pkg}/BUILD.bazel`, contents);
+}
+
+async function main() {
+  const packageJson = require(path.join(workspacePath, "package.json"));
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]: is a workspace always a package by definition? If yes, is it worth providing a better error message here? If not, how do we handle this?

File: internal/npm_install/generate_pnp_build_files.ts [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+    fs.mkdirSync(p);
+  }
+}
+
+function writeFileSync(p: string, content: string) {
+  mkdirp(path.dirname(p));
+  fs.writeFileSync(p, content);
+}
+
+function createBuildFile(pkg: string) {
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]:
```suggestion
function createBuildFile(packagePath: string) {
```
?

File: internal/npm_install/generate_pnp_build_files.ts (hunk 21 above)
Comment (joscha) [reply internal/npm_install/generate_pnp_build_files.ts]: or ideally call main from a CLI entry file.

File: internal/npm_install/process_pnp_file.ts [hunk 28]
```diff
@@ -0,0 +1,23 @@
+import * as fs from "fs";
+import * as process from "process";
+import * as path from "path";
+if (require.main === module) {
+  main();
```
Comment (joscha) [inline internal/npm_install/process_pnp_file.ts]:
```suggestion
  main(process.argv[3]);
```

File: internal/npm_install/process_pnp_file.ts [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+/**
+ * This patches a PnP file to force the resolver to resolve from the repository's directory
+ * not the entry_point's directory. PnP assumes they are the same so when bazel creates the
+ * pnp file in a different directory, the pnp files can't resolve the paths.
+ */
+function main() {
```
Comment (joscha) [inline internal/npm_install/process_pnp_file.ts]:
```suggestion
function main(workspacePath: string) {
```

File: internal/npm_install/process_pnp_file.ts [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+  const workspacePath = process.argv.slice(2)[0];
+
+  const pnpPath = path.join(workspacePath, ".pnp.js");
+  const pnpFile = fs.readFileSync(pnpPath, "utf8");
+  const patchedPnpFile = pnpFile.replace(
+    "issuerModule ? issuerModule.filename : `${process.cwd()}/`;",
+    `"${process.cwd()}"`
```
Comment (joscha) [inline internal/npm_install/process_pnp_file.ts]:
```suggestion
    JSON.stringify(process.cwd())
```

File: internal/npm_install/process_pnp_file.ts [hunk 31]
```diff
@@ -0,0 +1,23 @@
...
+ * This patches a PnP file to force the resolver to resolve from the repository's directory
+ * not the entry_point's directory. PnP assumes they are the same so when bazel creates the
+ * pnp file in a different directory, the pnp files can't resolve the paths.
+ */
+function main() {
+  const workspacePath = process.argv.slice(2)[0];
+
+  const pnpPath = path.join(workspacePath, ".pnp.js");
+  const pnpFile = fs.readFileSync(pnpPath, "utf8");
+  const patchedPnpFile = pnpFile.replace(
```
Comment (joscha) [inline internal/npm_install/process_pnp_file.ts:22]: can you add some docs here with an example of what this is supposed to match? Also, do we expect to only replace the first mach?

File: internal/npm_install/pre_process_package_json.js [hunk 32]
```diff
@@ -53,9 +53,13 @@ function main() {
     // Note: there is no equivalent npm functionality to clean out individual packages
     // from the npm cache.
     clearYarnFilePathCaches(pkg);
+    console.log(isPnp(pkg));
```
Comment (joscha) [inline internal/npm_install/pre_process_package_json.js]:
```suggestion
```

leftover?

File: internal/npm_install/generate_pnp_build_files.ts [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+)`;
+
+  writeFileSync(`${pkg}/BUILD.bazel`, contents);
+}
+
+async function main() {
+  const packageJson = require(path.join(workspacePath, "package.json"));
+  const dependencies = Object.keys(packageJson.dependencies || {});
+  const devDependencies = Object.keys(packageJson.dependencies || {});
+  dependencies.concat(devDependencies).forEach(createBuildFile);
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]: cheap enough, but there is probably no advantage over creating a new combination array and then iterating.

```
Object.keys(packageJson.dependencies || {}).forEach(createBuildFile);
Object.keys(packageJson.devDependencies || {}).forEach(createBuildFile);
```
will be as expressive but half the LOC.

or if you're keen:

```ts
[
  'dependencies',
  'devDependencies',
  'peerDependencies',
  'optionalDependencies'
]
  .forEach(d => Object.keys(packageJson[d] || {}).forEach(createBuildFile));
```

File: internal/npm_install/generate_pnp_build_files.ts (hunk 24 above)
Comment (danjuv) [reply internal/npm_install/generate_pnp_build_files.ts]: Not necessarily, but I didn't particularly want to deal with callbacks, but I see now fs has a promises property in node@^10, so it might be better to just use that

File: internal/npm_install/generate_pnp_build_files.ts (hunk 26 above)
Comment (danjuv) [reply internal/npm_install/generate_pnp_build_files.ts]: Yes, the workspacePath is the directory containing the package.json file. I suppose it's possible it doesn't exist, so we can return a better error message

File: internal/npm_install/process_pnp_file.ts (hunk 31 above)
Comment (danjuv) [reply internal/npm_install/process_pnp_file.ts:22]: we expect their to only ever be 1 match, this is a hack to force the path we look for PnP modules to be the same as where `yarn_install` is invoked. If we don't do this the pnp resolver defaults to the path the nodejs binary is invoked, which for bazel is different.

I suppose it's not necessarily only supposed to replace once, so we should do a global replace to make it better

File: internal/npm_install/pre_process_package_json.js (hunk 32 above)
Comment (danjuv) [reply internal/npm_install/pre_process_package_json.js]: This script is called by bazel as a way to tell the rule that the workspace is a pnp workspace, bazel can only take from stdout, so we need to log this to the console

File: internal/npm_install/generate_pnp_build_files.ts (hunk 33 above)
Comment (danjuv) [reply internal/npm_install/generate_pnp_build_files.ts]: second one is nice

File: internal/npm_install/generate_pnp_build_files.ts [hunk 34]
```diff
@@ -1,44 +1,47 @@
...
-  }
-}
-
 function writeFileSync(p: string, content: string) {
-  mkdirp(path.dirname(p));
-  fs.writeFileSync(p, content);
+  if (!existsSync(path.dirname(p))) {
+    mkdirSync(path.dirname(p), { recursive: true });
+  }
+  return fsp.writeFile(p, content);
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts:6]: this will return a Promise object now. Which is quite confusing because you use the sync methods above.

I suggest at least changing the signature to:
`function writeFile(p: string, content: string): Promise<void> {`

or even better:

```ts
async function writeFile(p: string, content: string): Promise<void> {
  await fsp.mkdir(path.dirname(p), { recursive: true });
  return await fsp.writeFile(p, content);
}
```

File: internal/npm_install/generate_pnp_build_files.ts [hunk 35]
```diff
@@ -1,44 +1,47 @@
...
 filegroup(
-    name = "${pkg}",
+    name = "${packagePath}",
     srcs = [] # we dont care about the actual content, we want a list of dependencies to modify the pnp file around
-)`;
+)
+`;
 
-  writeFileSync(`${pkg}/BUILD.bazel`, contents);
+  return writeFileSync(`${packagePath}/BUILD.bazel`, contents);
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]: writeFile is not synchronous anymore which makes it return a promise and thus also makes `createBuildFile` async.

File: internal/npm_install/generate_pnp_build_files.ts [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+])`;
+  const packageJson = require(path.join(workspacePath, "package.json"));
+  await Promise.all(
+    [
+      "dependencies",
+      "devDependencies",
+      "peerDependencies",
+      "optionalDependencies"
+    ].map(d => {
+      Promise.all(
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]: the value of this is never returned up from the map, so if anything gets rejected it will be unhandled. It also means that your outer Promise.all waits for nothing:

```suggestion
      return Promise.all(
```

File: internal/npm_install/generate_pnp_build_files.ts [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+        Object.keys(packageJson[d] || {})
+          .map(createBuildFile)
+          .concat([writeFileSync(`BUILD.bazel`, contents)])
```
Comment (joscha) [inline internal/npm_install/generate_pnp_build_files.ts]:
```suggestion
          .concat([writeFileSync("BUILD.bazel", contents)])
```
or
```suggestion
          .concat([writeFileSync('BUILD.bazel', contents)])
```

as you're not using any interpolation.

File: internal/npm_install/pre_process_package_json.js [hunk 38]
```diff
@@ -53,9 +53,13 @@ function main() {
     // Note: there is no equivalent npm functionality to clean out individual packages
     // from the npm cache.
     clearYarnFilePathCaches(pkg);
+    console.log(JSON.stringify({pnp: isPnp(pkg)}));
```
Comment (joscha) [inline internal/npm_install/pre_process_package_json.js]:
```suggestion
    process.stdout.write(JSON.stringify({pnp: isPnp(pkg)}));
```
makes it more obvious that this is a deliberate print to sdtout.

File: internal/npm_install/generate_pnp_build_files.ts (hunk 34 above)
Comment (danjuv) [reply internal/npm_install/generate_pnp_build_files.ts:6]: yeah, this should be renamed

File: internal/npm_install/generate_pnp_build_files.ts (hunk 35 above)
Comment (danjuv) [reply internal/npm_install/generate_pnp_build_files.ts]: this is intentional, I should be dealing with all the async stuff with `Promise.all` though I may have missed something

File: third_party/github.com/bazel_json/LICENSE [hunk 39]
```diff
@@ -0,0 +1,20 @@
+Copyright 2019 Erick Johnson <ejohnson82@gmail.com>
```
Comment (danjuv) [inline third_party/github.com/bazel_json/LICENSE:1]: third party json parser, we can't introduce it as an http_archive since you can't patch the workspace file in a repo

Comment (uri-canva) [reply third_party/github.com/bazel_json/LICENSE:1]: I don't think that fully explains why the vendoring is necessary, but there is an example of this upstream already so I think this is ok:
https://github.com/bazelbuild/rules_nodejs/tree/a13f2b6e2a132954fb63b7a905818cf8a6ceb824/third_party/github.com/bazelbuild/bazel-skylib/rules

File: internal/npm_install/npm_install.bzl (hunk 18 above)
Comment (danjuv) [reply internal/npm_install/npm_install.bzl:143]: I had a go at this, it seems to introduce too many merge conflicts to the point that I don't think it's too worthwhile

File: internal/node/node.bzl [hunk 40]
```diff
@@ -200,7 +200,7 @@ def _nodejs_binary_impl(ctx):
     node_tool_files.append(ctx.file._bazel_require_script)
 
     if not ctx.outputs.templated_args_file:
-        templated_args = ctx.attr.templated_args
+        templated_args = [a for a in ctx.attr.templated_args]
```
Comment (danjuv) [inline internal/node/node.bzl:203]: this string list is frozen, so I created a new list so we can mutate it (add the pnp flag if necessary)

File: internal/npm_install/npm_install.bzl (hunk 18 above)
Comment (uri-canva) [reply internal/npm_install/npm_install.bzl:143]: This will be necessary once we apply the patches in the repo anyway: we don't want to specify a single patch that contains all the changes we made to upstream, we want a separate patch for each logical change, so we know when we can stop patching them in because upstream have merged our PRs, or because we don't need them anymore.

Comment (danjuv) [reply internal/npm_install/npm_install.bzl:143]: good point, i didnt think about it that way, i'll split this out

File: internal/node/node.bzl (hunk 40 above)
Comment (uri-canva) [inline internal/node/node.bzl:203]: Try using `.to_list()`, I'm not 100% sure it works, but if it does, it's nicer than an identity list comprehension.

File: internal/node/node.bzl [hunk 41]
```diff
@@ -226,6 +226,10 @@ def _nodejs_binary_impl(ctx):
 
     is_builtin = ctx.attr._node.label.workspace_name in ["nodejs_%s" % p for p in BUILT_IN_NODE_PLATFORMS]
 
+    output_files = [ctx.outputs.loader]
```
Comment (uri-canva) [inline internal/node/node.bzl:229]: Why is this called `output_files`? Isn't `ctx.file.pnp_file` an input file?

File: internal/npm_install/npm_install.bzl [hunk 42]
```diff
@@ -277,6 +310,7 @@ cd "{root}" && "{npm}" {npm_args}
     )
     if result.return_code:
         fail("pre_process_package_json.js failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (result.stdout, result.stderr))
+    is_pnp = json_parse(result.stdout)["pnp"]
```
Comment (uri-canva) [inline internal/npm_install/npm_install.bzl:313]: Since you can parse json, why not parse `package.json` directly, rather than going through the script?

**Review by uri-canva — APPROVED** [review]
Very close! I think if we can read the `package.json` from starlark then it's good.

File: internal/npm_install/npm_install.bzl (hunk 42 above)
Comment (danjuv) [reply internal/npm_install/npm_install.bzl:313]: Because Starlark doesn't support file i/o as far as I know

Comment (uri-canva) [reply internal/npm_install/npm_install.bzl:313]: Try https://docs.bazel.build/versions/master/skylark/lib/repository_ctx.html#read.

Comment (danjuv) [reply internal/npm_install/npm_install.bzl:313]: oh nice, makes sense it's exposed on repository rules

**Comment by danjuv** [comment]
see https://canvadev.atlassian.net/browse/DEVF-831 for more details
